gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/dsl/graph/GraphTraversal.java

```diff
@@ -2554,21 +2557,30 @@ public default GraphTraversal<S, E> property(final VertexProperty.Cardinality ca
      */
     public default GraphTraversal<S, E> property(final Object key, final Object value, final Object... keyValues) {
         if (key instanceof VertexProperty.Cardinality) {
-            if (value instanceof Map) { //Handle the property(Cardinality, Map) signature
-                final Map<Object, Object> map = (Map)value;
+            if (value instanceof Map) {
+                // Handle the property(Cardinality, Map) signature
+                final Map<Object, Object> map = (Map) value;
                 for (Map.Entry<Object, Object> entry : map.entrySet()) {
-                    property(key, entry.getKey(), entry.getValue());
+                    final Object val = entry.getValue();
+                    if (val instanceof CardinalityValueTraversal) {
+                        final CardinalityValueTraversal cardVal = (CardinalityValueTraversal) val;
+                        property(cardVal.getCardinality(), entry.getKey(), cardVal.getValue());
+                    } else {
+                        property(key, entry.getKey(), entry.getValue());
```

Comment:
probably recursive call is not safe here
```
gremlin> g.V(1).property(list, [["k1":"v1", "k2":"v2"]:"value"])
==>v[1]
gremlin> g.V(1).valueMap()
==>[k1:[v1],k2:[v2],name:[marko],age:[29]]
```

Reply:
hmm, that's right. it seems that has always been a problem. changing it to prevent traversal construction if you provide that sort of input.